File: pharmacy-ecommerce/scripts/import_products.py
#!/usr/bin/env python3
"""
Script para importar productos desde Excel a PostgreSQL
Actualizado para el nuevo formato de Excel 2026-01-19_LISTA_DE_PRECIOS.xlsx
Columnas: ID, PRODUCTO, LABORATORIO, DEPARTAMENTO, ACCIÓN TERAPÉUTICA, PRINCIPIO ACTIVO,
          UNIDADES PRESENTACIÓN, PRESENTACIÓN, RECETA MÉDICA, CONTROL LEGAL, ES BIOEQUIVALENTE,
          REGISTRO SANITARIO, TITULAR REGISTRO, STOCK, PRECIO, PRECIO POR UNIDAD
"""

import pandas as pd
import psycopg2
import re
import unicodedata
import os

# Database connection - use environment variable or default to Railway
DB_URL = os.environ.get(
    'DATABASE_URL',
    'postgresql://postgres:RLfkDsoXZmYIjkeoavlWkNXaCGFovWXT@maglev.proxy.rlwy.net:24761/railway'
)

def slugify(text):
    """Convert text to URL-friendly slug"""
    if not text or pd.isna(text):
        return ""
    # Normalize unicode characters
    text = str(text).lower()
    text = unicodedata.normalize('NFKD', text)
    text = text.encode('ascii', 'ignore').decode('ascii')
    # Replace spaces and special chars with hyphens
    text = re.sub(r'[^a-z0-9]+', '-', text)
    # Remove leading/trailing hyphens
    text = text.strip('-')
    return text[:100]  # Limit length

def clean_text(text):
    """Clean text by removing special characters like degree symbol"""
    if not text or pd.isna(text):
        return ""
    text = str(text).strip()
    # Replace degree symbol and other problematic chars
    text = text.replace('°', '').replace('º', '')
    # Normalize unicode
    text = unicodedata.normalize('NFKC', text)
    return text

def build_description(row):
    """Build a rich description from product details"""
    parts = []

    if row.get('accion_terapeutica') and not pd.isna(row['accion_terapeutica']):
        parts.append(f"Acción terapéutica: {clean_text(row['accion_terapeutica'])}")

    if row.get('principio_activo') and not pd.isna(row['principio_activo']):
        parts.append(f"Principio activo: {clean_text(row['principio_activo'])}")

    if row.get('presentacion') and not pd.isna(row['presentacion']):
        parts.append(f"Presentación: {clean_text(row['presentacion'])}")

    if row.get('receta') and not pd.isna(row['receta']):
        receta = clean_text(row['receta'])
        if receta:
            parts.append(f"Requiere: {receta}")

    return ". ".join(parts) if parts else f"Producto farmacéutico: {clean_text(row.get('producto', ''))}"

def main():
    # Determine script directory to find Excel file
    script_dir = os.path.dirname(os.path.abspath(__file__))
    parent_dir = os.path.dirname(script_dir)
    excel_path = os.path.join(parent_dir, '2026-01-19_LISTA_DE_PRECIOS.xlsx')

    # Read Excel
    print(f"Leyendo Excel desde: {excel_path}")
    df = pd.read_excel(excel_path, header=None)

    # Skip header row (first row is header)
    df = df.iloc[1:]
    df.columns = [
        'id', 'producto', 'laboratorio', 'departamento', 'accion_terapeutica',
        'principio_activo', 'unidades_presentacion', 'presentacion', 'receta',
        'control_legal', 'es_bioequivalente', 'registro_sanitario',
        'titular_registro', 'stock', 'precio', 'precio_por_unidad'
    ]

    # Clean data
    df = df.dropna(subset=['producto'])  # Remove rows without product name
    df['stock'] = pd.to_numeric(df['stock'], errors='coerce').fillna(0).astype(int)
    df['precio'] = pd.to_numeric(df['precio'], errors='coerce').fillna(0)

    # Clean text fields
    df['producto'] = df['producto'].apply(clean_text)
    df['laboratorio'] = df['laboratorio'].apply(clean_text)
    df['departamento'] = df['departamento'].apply(clean_text)
    df['id'] = df['id'].apply(lambda x: str(x).strip() if not pd.isna(x) else '')

    # Filter only products with invalid price (price = 0 or negative)
    # Include products with stock = 0 (they will show as "Agotado")
    df = df[df['precio'] > 0]

    print(f"Productos validos para importar: {len(df)}")
    print(f"  - Con stock > 0: {len(df[df['stock'] > 0])}")
    print(f"  - Con stock = 0 (Agotados): {len(df[df['stock'] == 0])}")

    # Connect to database
    print(f"\nConectando a la base de datos...")
    conn = psycopg2.connect(DB_URL)
    cur = conn.cursor()

    # Get existing categories or create new ones based on DEPARTAMENTO
    print("Procesando categorias (departamentos)...")
    categories = df['departamento'].dropna().unique()
    category_map = {}

    for cat_name in categories:
        if not cat_name or pd.isna(cat_name):
            continue

        cat_name = str(cat_name).strip()
        if not cat_name:
            continue

        slug = slugify(cat_name)
        if not slug:
            continue

        # Check if category exists
        cur.execute("SELECT id FROM categories WHERE slug = %s", (slug,))
        result = cur.fetchone()

        if result:
            category_map[cat_name] = result[0]
        else:
            # Create category
            cur.execute(
                "INSERT INTO categories (name, slug, description, active) VALUES (%s, %s, %s, %s) RETURNING id",
                (cat_name, slug, f"Productos de {cat_name}", True)
            )
            category_map[cat_name] = cur.fetchone()[0]

    conn.commit()
    print(f"Categorias procesadas: {len(category_map)}")

    # Delete existing products (user requested full replacement)
    print("\nLimpiando productos anteriores...")
    cur.execute("DELETE FROM order_items")
    cur.execute("DELETE FROM products")
    conn.commit()
    print("Productos eliminados.")

    # Insert products
    print("\nInsertando productos...")
    products_inserted = 0
    products_with_stock = 0
    products_out_of_stock = 0

    for _, row in df.iterrows():
        try:
            name = str(row['producto']).strip()
            if not name:
                continue

            slug = slugify(name)
            if not slug:
                continue

            # Make slug unique by adding counter if needed
            base_slug = slug
            counter = 1
            while True:
                cur.execute("SELECT id FROM products WHERE slug = %s", (slug,))
                if not cur.fetchone():
                    break
                slug = f"{base_slug}-{counter}"
                counter += 1

            stock = int(row['stock']) if not pd.isna(row['stock']) else 0
            price = float(row['precio']) if not pd.isna(row['precio']) else 0
            external_id = str(row['id']).strip() if not pd.isna(row['id']) else None
            laboratory = str(row['laboratorio']).strip() if not pd.isna(row['laboratorio']) else None

            # Get category from DEPARTAMENTO
            cat_name = str(row['departamento']).strip() if not pd.isna(row['departamento']) else None
            category_id = category_map.get(cat_name) if cat_name else None

            # Build rich description
            description = build_description(row)

            # Insert product with new fields
            cur.execute("""
                INSERT INTO products (name, slug, description, price, stock, category_id, external_id, laboratory, active)
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)
            """, (
                name,
                slug,
                description,
                price,
                stock,
                category_id,
                external_id,
                laboratory,
                True
            ))

            products_inserted += 1
            if stock > 0:
                products_with_stock += 1
            else:
                products_out_of_stock += 1

            if products_inserted % 100 == 0:
                print(f"  Insertados: {products_inserted}")
                conn.commit()

        except Exception as e:
            print(f"Error en producto {row['producto']}: {e}")
            continue

    conn.commit()

    # Summary
    cur.execute("SELECT COUNT(*) FROM products")
    total_products = cur.fetchone()[0]

    cur.execute("SELECT COUNT(*) FROM products WHERE stock > 0")
    available_products = cur.fetchone()[0]

    cur.execute("SELECT COUNT(*) FROM products WHERE stock = 0")
    out_of_stock_products = cur.fetchone()[0]

    cur.execute("SELECT COUNT(*) FROM categories")
    total_categories = cur.fetchone()[0]

    cur.execute("SELECT SUM(stock * price) FROM products")
    inventory_value = cur.fetchone()[0] or 0

    print(f"\n{'='*50}")
    print(f"=== IMPORTACION COMPLETADA ===")
    print(f"{'='*50}")
    print(f"Productos importados: {products_inserted}")
    print(f"  - Con stock disponible: {products_with_stock}")
    print(f"  - Agotados (stock = 0): {products_out_of_stock}")
    print(f"Total productos en DB: {total_products}")
    print(f"Total categorias: {total_categories}")
    print(f"Valor total inventario: ${inventory_value:,.0f} CLP")
    print(f"{'='*50}")

    cur.close()
    conn.close()

if __name__ == "__main__":
    main()

File: pharmacy-ecommerce/scripts/import_products.js
#!/usr/bin/env node
/**
 * Script para importar productos desde Excel a PostgreSQL
 * Versión Node.js del script Python
 */

const xlsx = require('xlsx');
const { Client } = require('pg');
const path = require('path');

// Database connection
const DB_URL = process.env.DATABASE_URL ||
  'postgresql://postgres:RLfkDsoXZmYIjkeoavlWkNXaCGFovWXT@maglev.proxy.rlwy.net:24761/railway';

function slugify(text) {
  if (!text) return '';
  return text
    .toString()
    .toLowerCase()
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '') // Remove accents
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
    .slice(0, 100);
}

function cleanText(text) {
  if (!text) return '';
  return text.toString().trim().replace(/°|º/g, '');
}

function buildDescription(row) {
  const parts = [];

  if (row.accion_terapeutica) {
    parts.push(`Acción terapéutica: ${cleanText(row.accion_terapeutica)}`);
  }
  if (row.principio_activo) {
    parts.push(`Principio activo: ${cleanText(row.principio_activo)}`);
  }
  if (row.presentacion) {
    parts.push(`Presentación: ${cleanText(row.presentacion)}`);
  }
  if (row.receta) {
    parts.push(`Requiere: ${cleanText(row.receta)}`);
  }

  return parts.length > 0 ? parts.join('. ') : `Producto farmacéutico: ${cleanText(row.producto)}`;
}

async function main() {
  const excelPath = path.join(__dirname, '..', '2026-01-19_LISTA_DE_PRECIOS.xlsx');

  console.log(`Leyendo Excel desde: ${excelPath}`);

  const workbook = xlsx.readFile(excelPath);
  const sheetName = workbook.SheetNames[0];
  const worksheet = workbook.Sheets[sheetName];
  const rawData = xlsx.utils.sheet_to_json(worksheet, { header: 1, defval: '' });

  // Skip header row
  const headers = ['id', 'producto', 'laboratorio', 'departamento', 'accion_terapeutica',
    'principio_activo', 'unidades_presentacion', 'presentacion', 'receta',
    'control_legal', 'es_bioequivalente', 'registro_sanitario',
    'titular_registro', 'stock', 'precio', 'precio_por_unidad'];

  const data = rawData.slice(1).map(row => {
    const obj = {};
    headers.forEach((h, i) => obj[h] = row[i]);
    return obj;
  });

  // Filter valid products
  const products = data.filter(p => {
    const producto = cleanText(p.producto);
    const precio = parseFloat(p.precio) || 0;
    return producto && precio > 0;
  });

  const withStock = products.filter(p => (parseInt(p.stock) || 0) > 0).length;
  const outOfStock = products.filter(p => (parseInt(p.stock) || 0) === 0).length;

  console.log(`Productos válidos para importar: ${products.length}`);
  console.log(`  - Con stock > 0: ${withStock}`);
  console.log(`  - Con stock = 0 (Agotados): ${outOfStock}`);

  // Connect to database
  console.log('\nConectando a la base de datos...');
  const client = new Client({ connectionString: DB_URL });
  await client.connect();

  // Process categories (departamentos)
  console.log('Procesando categorías (departamentos)...');
  const departamentos = [...new Set(products.map(p => cleanText(p.departamento)).filter(Boolean))];
  const categoryMap = {};

  for (const catName of departamentos) {
    const slug = slugify(catName);
    if (!slug) continue;

    const result = await client.query('SELECT id FROM categories WHERE slug = $1', [slug]);

    if (result.rows.length > 0) {
      categoryMap[catName] = result.rows[0].id;
    } else {
      const insert = await client.query(
        'INSERT INTO categories (name, slug, description, active) VALUES ($1, $2, $3, $4) RETURNING id',
        [catName, slug, `Productos de ${catName}`, true]
      );
      categoryMap[catName] = insert.rows[0].id;
    }
  }
  console.log(`Categorías procesadas: ${Object.keys(categoryMap).length}`);

  // Delete existing products
  console.log('\nLimpiando productos anteriores...');
  await client.query('DELETE FROM order_items');
  await client.query('DELETE FROM products');
  console.log('Productos eliminados.');

  // Insert products
  console.log('\nInsertando productos...');
  let productsInserted = 0;
  let productsWithStock = 0;
  let productsOutOfStock = 0;
  const usedSlugs = new Set();

  for (const row of products) {
    try {
      const name = cleanText(row.producto);
      if (!name) continue;

      let slug = slugify(name);
      if (!slug) continue;

      // Make slug unique
      let baseSlug = slug;
      let counter = 1;
      while (usedSlugs.has(slug)) {
        slug = `${baseSlug}-${counter}`;
        counter++;
      }
      usedSlugs.add(slug);

      const stock = parseInt(row.stock) || 0;
      const price = parseFloat(row.precio) || 0;
      const externalId = row.id ? row.id.toString().trim() : null;
      const laboratory = cleanText(row.laboratorio) || null;
      const catName = cleanText(row.departamento);
      const categoryId = categoryMap[catName] || null;
      const description = buildDescription(row);

      await client.query(`
        INSERT INTO products (name, slug, description, price, stock, category_id, external_id, laboratory, active)
        VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
      `, [name, slug, description, price, stock, categoryId, externalId, laboratory, true]);

      productsInserted++;
      if (stock > 0) productsWithStock++;
      else productsOutOfStock++;

      if (productsInserted % 100 === 0) {
        console.log(`  Insertados: ${productsInserted}`);
      }
    } catch (error) {
      console.error(`Error en producto ${row.producto}: ${error.message}`);
    }
  }

  // Summary
  const totalRes = await client.query('SELECT COUNT(*) FROM products');
  const availableRes = await client.query('SELECT COUNT(*) FROM products WHERE stock > 0');
  const outOfStockRes = await client.query('SELECT COUNT(*) FROM products WHERE stock = 0');
  const categoriesRes = await client.query('SELECT COUNT(*) FROM categories');
  const valueRes = await client.query('SELECT SUM(stock * price) FROM products');

  console.log('\n' + '='.repeat(50));
  console.log('=== IMPORTACIÓN COMPLETADA ===');
  console.log('='.repeat(50));
  console.log(`Productos importados: ${productsInserted}`);
  console.log(`  - Con stock disponible: ${productsWithStock}`);
  console.log(`  - Agotados (stock = 0): ${productsOutOfStock}`);
  console.log(`Total productos en DB: ${totalRes.rows[0].count}`);
  console.log(`Total categorías: ${categoriesRes.rows[0].count}`);
  console.log(`Valor total inventario: $${Math.round(valueRes.rows[0].sum || 0).toLocaleString()} CLP`);
  console.log('='.repeat(50));

  await client.end();
}

main().catch(console.error);
